Fix implicit to and from bool type juggling

Girgias opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 7.4 (but can rebase to lower)
| Bug fix?      | maybe
| New feature?  | no
| Deprecations? | no
| Issues        | N/A
| License       | MIT

This is to see part of the impact of https://github.com/php/php-src/pull/18879, and it seems to be finding some bugs in tests or SF itself, but I would need confirmation from other people.

This is very much W.I.P. and I might use CI to check that my changes are correct as I have some local test failures even just running with 8.4.8

I am well aware the commit messages are kinda useless at the moment, but they are somewhat self-contained, so they can be looked at individually.

Girgias committed:
Use false instead of 0

Girgias committed:
IntlTestHelper::requireFullIntl() use null as second arg instead of false

Girgias committed:
Check return value of getenv()

Girgias committed:
Symfony\Component\Validator\Constraints\Isbn does not take a bool as first arg

Girgias committed:
Cast result of preg_match to bool

Girgias committed:
Use bool as param type

Girgias committed:
Cast bitwise op result to bool

Girgias committed:
ErrorHandler::handleError() cast bitwise op to bool and use false instead of 0

Girgias committed:
Pull fixes from Nicolas

Girgias committed:
Pull ignores/suppressed tests from Nicolas

Girgias committed:
Yield empty string when no doc comment exists

stof commented on src/Symfony/Component/DependencyInjection/Loader/YamlFileLoader.php:
```diff
@@ -568,7 +568,7 @@ private function parseDefinition(string $id, array|string|null $service, string
                     $method = $call['method'];
                     $args = $call['arguments'] ?? [];
-                    $returnsClone = $call['returns_clone'] ?? false;
+                    $returnsClone = $call['returns_clone'] ? true : false;
```
this will break if `returns_clone` is not defined in the array.

and that's a case where we expect a boolean if the key is defined, or an undefined key (which should then default to `false`)

Girgias replied:
Indeed this seems to be not necessary

Girgias committed:
Set value to null if parse_url() fails badly and returns false

stof commented on src/Symfony/Component/DependencyInjection/Definition.php:
```diff
@@ -307,7 +307,7 @@ public function setMethodCalls(array $calls = []): static
         $this->calls = [];
         foreach ($calls as $call) {
-            $this->addMethodCall($call[0], $call[1], $call[2] ?? false);
+            $this->addMethodCall($call[0], $call[1], $call[2] ? true : false);
```
this is a case where the third value in the array is expected to be a boolean or absent, which is broken in this change.

Girgias replied:
Indeed this seems to be not necessary

Girgias committed:
Fix deprecation with PHPUnit's assertTrue

Not sure what those asserts are doing however

stof commented on src/Symfony/Component/DependencyInjection/Loader/XmlFileLoader.php:
```diff
@@ -339,7 +339,11 @@ private function parseDefinition(\DOMElement $service, string $file, Definition
+            $definition->addMethodCall(
+                $call->getAttribute('method'),
+                $this->getArgumentsAsPhp($call, 'argument', $file),
+                (bool) XmlUtils::phpize($call->getAttribute('returns-clone'))
```
I suggest doing a method handling of the case of the missing attribute (instead of handling it as `$call->getAttribute('returns-clone')` returning an empty string which then casts to `false`)

Girgias replied:
Do you want me to basically just do:
```suggestion
                XmlUtils::phpize($call->getAttribute('returns-clone')) !== ''
```
or something else? Because I don't really understand your suggestion.

stof replied:
Doing something like `$call->getAttribute('returns-clone') !== '' ? XmlUtils::phpize($call->getAttribute('returns-clone')) : false`, as we need to detect the case of the omitted attribute (to default to `false`) while still using the proper output of `phpize` for cases using the attribute.

Girgias replied:
Done

stof commented on src/Symfony/Component/DependencyInjection/Loader/YamlFileLoader.php:
```diff
@@ -589,7 +589,7 @@ private function parseDefinition(string $id, array|string|null $service, string
                         $method = $call[0];
                         $args = $call[1] ?? [];
-                        $returnsClone = $call[2] ?? false;
+                        $returnsClone = $call[2] ? true : false;
```
the existing code should probably be kept here as well.

Girgias committed:
Remove ZPP test

There is no need to test false as this is a PHP engine behaviour to cast to empty string

Girgias committed:
file_get_contents() does not take a fopen mode parameter

Girgias committed:
Symfony\Component\String\LazyString::resolve() add explicit string cast

Girgias committed:
Fix string to true type juggling in test

stof commented on src/Symfony/Component/ErrorHandler/ErrorHandler.php:
```diff
@@ -402,7 +402,7 @@ public function handleError(int $type, string $message, string $file, int $line)
         // Never throw on warnings triggered by assert()
         if (\E_WARNING === $type && 'a' === $message[0] && 0 === strncmp($message, 'assert(): ', 10)) {
-            $throw = 0;
+            $throw = false;
```
`$throw` is still an int if we don't enter this `if` branch, as it is defined with bitwise comparisons on line 400. It needs to be fixed there as well.

Girgias committed:
Remove ZPP test

There is no need to test false as this is a PHP engine behaviour to cast to empty string

Girgias committed:
Use empty string instead of false

stof commented on src/Symfony/Component/DependencyInjection/Tests/Compiler/CheckTypeDeclarationsPassTest.php:
```diff
@@ -336,7 +336,7 @@ public function testProcessFailsOnPassingClassToScalarTypedParameter()
     }
 
-    public function testProcessSuccessOnPassingBadScalarType()
+    public function xtestProcessSuccessOnPassingBadScalarType()
```
why skipping this test ?

Girgias replied:
I stole this from @nicolas-grekas WIP commit: https://github.com/nicolas-grekas/symfony/commit/02e3881f4d6b8f6059d1514da8704f0890d5e632

nicolas-grekas replied:
:see_no_evil:

stof commented on src/Symfony/Bundle/TwigBundle/Tests/Functional/AttributeExtensionTest.php:
```diff
@@ -51,7 +51,7 @@ public function registerContainerConfiguration(LoaderInterface $loader): void
                         ->setAutoconfigured(true);
                     $container->register(RuntimeExtensionWithAttributes::class, RuntimeExtensionWithAttributes::class)
-                        ->setArguments(['prefix_'])
+                        ->setArguments([true])
```
given that `RuntimeExtensionWithAttributes` performs string concatenation with the prefix, the right fix is rather to change the constructor argument type in this RuntimeExtensionWithAttributes stub class.

stof commented on src/Symfony/Component/String/LazyString.php:
```diff
@@ -76,7 +76,8 @@ final public static function isStringable(mixed $value): bool
     final public static function resolve(\Stringable|string|int|float|bool $value): string
     {
-        return $value;
+        // TODO Do something more useful for false and true?
```
the TODO should be removed, as a string cast is also what we do in `fromStringable`

Girgias committed:
intl false instead of null

Girgias committed:
Explicit cast to bool from string

Girgias committed:
Definition::addMethodCall() arg 3 must be a bool v2

stof commented on src/Symfony/Component/DependencyInjection/Definition.php:
```diff
@@ -307,7 +307,7 @@ public function setMethodCalls(array $calls = []): static
         $this->calls = [];
         foreach ($calls as $call) {
-            $this->addMethodCall($call[0], $call[1], $call[2] ?? false);
+            $this->addMethodCall($call[0], $call[1], isset($call[2]) && $call[2]);
```
Do you have stack trace for the case triggering the deprecation here (without your patch) ? Maybe we have some tests not using the expected structure for the array, but it might be better to fix the root cause instead of a symptom (replacing the boolean type juggling in function context by a boolean type juggling in boolean operator context, which is not deprecated)

stof commented on src/Symfony/Component/DependencyInjection/Loader/YamlFileLoader.php:
```diff
@@ -589,7 +589,7 @@ private function parseDefinition(string $id, array|string|null $service, string
                         $method = $call[0];
                         $args = $call[1] ?? [];
-                        $returnsClone = $call[2] ?? false;
+                        $returnsClone = isset($call[2]) && $call[2];
```
Same for those 2 cases. My expectation here is that we always expected a boolean Yaml value (or nothing) in such case, which would not trigger the deprecation with the existing code.
If we have some tests using non-boolean values, we might want to fix those tests instead (and maybe trigger a deprecation in Symfony for projects using a non-boolean value).

stof commented on src/Symfony/Component/Form/Tests/Extension/Core/DataTransformer/NumberToLocalizedStringTransformerTest.php:
```diff
@@ -65,7 +65,7 @@ public static function provideTransformations()
     {
         // Since we test against other locales, we need the full implementation
-        IntlTestHelper::requireFullIntl($this, false);
+        IntlTestHelper::requireFullIntl($this, null);
```
I suggest omitting `null` entirely as it is optional (as done in some of the impacted files btw).

Girgias replied:
I'll do that when I have time and have gone through the other issues.

Girgias committed:
Fix test passing false to ?string type

Girgias committed:
Remove ZPP tests testing false to string coercion

Girgias committed:
NumberFormatter::setAttribute() takes values of type int|float

This is possibly something that should be fixed on PHP's side

Girgias commented:
There are a few cases where I don't seem to be able to find the root cause. A bit one is:
```
  1062x: Symfony\Component\Console\Tests\Descriptor\AbstractDescriptorTestCase::normalizeOutput(): Passing bool to parameter #1 ($output) of type string is deprecated
    942x in JsonDescriptorTest::testDescribeApplication from Symfony\Component\Console\Tests\Descriptor
    56x in JsonDescriptorTest::testDescribeInputOption from Symfony\Component\Console\Tests\Descriptor
    24x in JsonDescriptorTest::testDescribeInputArgument from Symfony\Component\Console\Tests\Descriptor
    24x in JsonDescriptorTest::testDescribeInputDefinition from Symfony\Component\Console\Tests\Descriptor
    16x in JsonDescriptorTest::testDescribeCommand from Symfony\Component\Console\Tests\Descriptor
```

Girgias committed:
Default non-existing environment variable to empty string

Girgias committed:
Split grapheme extraction and check if one is found

Girgias committed:
Fix substr with bool?

Girgias committed:
Use '1' instead of true

Girgias committed:
cookie_httponly key does not support values of auto

stof commented:
@Girgias you can define the `SYMFONY_DEPRECATIONS_HELPER` env variable with a regex using `/` as delimiter (this is important as only strings starting with `/` are considered to be regex in the different syntaxes supported for that config). When this regex matches the message of a deprecation, the test execution will stop and the backtrace of the deprecation will be reported. This allows to identify the source of the deprecation.
